### PR TITLE
Update syntax for Julia v1

### PR DESCRIPTION
**Pull request by nickrobinson251**
- Is `ReentrantLock` what we want here?
- I couldn't find the JuliaLang/Julia PR that moved `AbstractLock` out of the `Threads` submodule
- I didn't touch appveyor... we proabably want to update that (https://github.com/invenia/ReadWriteLocks.jl/pull/2)
- step one of https://github.com/invenia/ReadWriteLocks.jl/issues/3

### CODE REVIEW

**Comment by iamed2** [comment]
The commit that moved it out was https://github.com/JuliaLang/julia/commit/a0b7a7610334d0925f126ae20f5ff2d074e3ba69

**Comment by iamed2** [comment]
Try removing appveyor and using Travis Windows (just add `windows` as an OS)

File: test/runtests.jl [hunk 1]
```diff
@@ -1,2 +1,6 @@
-# write your own tests here
-include("singlethreaded.jl")
+using ReadWriteLocks
+using Test
+
+@testset "ReadWriteLocks" begin
+    include("singlethreaded.jl")
```
Comment (iamed2) [inline test/runtests.jl]: What about the `"Two-threaded tests"` testset?

File: src/ReadWriteLocks.jl [hunk 2]
```diff
@@ -49,7 +39,7 @@ write_lock(rwlock::ReadWriteLock) = rwlock.write_lock
 
 function lock!(read_lock::ReadLock)
```
Comment (iamed2) [inline src/ReadWriteLocks.jl:47]: This API should be changed to match the `AbstractLock` API, including the other API functions like `trylock`.

File: src/ReadWriteLocks.jl [hunk 3]
```diff
@@ -1,42 +1,32 @@
...
+struct WriteLock{T<:AbstractLock}
     rwlock::T
 end
 
-type ReadWriteLock <: _Lock
+mutable struct ReadWriteLock <: AbstractLock
     readers::Int
     writer::Bool
-    lock::Mutex  # reentrant mutex
+    lock::ReentrantLock  # reentrant mutex
```
Comment (iamed2) [inline src/ReadWriteLocks.jl]: I think we should make ReadWriteLock parameterized by an `AbstractLock` type, and just make `ReentrantLock` the default.

File: test/runtests.jl (hunk 1 above)
Comment (nickrobinson251) [reply test/runtests.jl]: you named it 😆 I will just move them into the runtests file 😄

File: test/singlethreaded.jl [hunk 4]
```diff
@@ -1,120 +1,121 @@
...
 
             @async begin
-                @fact rwlock.writer --> false
-                @fact rwlock.readers --> 0
+                @test rwlock.writer == false
+                @test rwlock.readers == 0
                 for i = 1:NUM_LOCKS
-                    @fact lock!(rlock) --> nothing
-                    @fact rwlock.readers --> i
+                    @test lock!(rlock) == nothing
```
Comment (iamed2) [inline test/singlethreaded.jl]: We don't need to do any tests of the return values of `lock` and `unlock` anymore. We can just call the functions.

FactCheck was not as good as Test at handling errors outside of tests.

File: test/runtests.jl [hunk 5]
```diff
@@ -1,2 +1,6 @@
-# write your own tests here
-include("singlethreaded.jl")
+using ReadWriteLocks
+using Test
+
+@testset "ReadWriteLocks" begin
```
Comment (iamed2) [inline test/runtests.jl]: Now that Julia 1.2 and 1.3 exist, we should add tests using actual threads. We can talk about that on Slack.

**Review by iamed2 — CHANGES_REQUESTED** [review]

File: src/ReadWriteLocks.jl (hunk 3 above)
Comment (nickrobinson251) [reply src/ReadWriteLocks.jl]: one slight hiccup is we require v1.2 to have `ReentrantLock <: Base.AbstractLock`

Comment (iamed2) [reply src/ReadWriteLocks.jl]: A `Union` is fine

File: test/runtests.jl (hunk 1 above)
Comment (iamed2) [reply test/runtests.jl]: Oh lol, maybe because they were never actually run on multiple threads?

File: test/singlethreaded.jl (hunk 4 above)
Comment (nickrobinson251) [reply test/singlethreaded.jl]: hmm, changing all cases of `@test [un]lock!(x) == nothing` -> `[un]lock!(x)` causes tests to fail -- we hit code we were not hitting before 🤔

Comment (iamed2) [reply test/singlethreaded.jl]: Push a commit that does that so I can see

File: src/ReadWriteLocks.jl [hunk 6]
```diff
@@ -1,42 +1,39 @@
...
 end
 
-immutable WriteLock{T<:_Lock}
+struct WriteLock{T<:AbstractLock}
     rwlock::T
 end
 
-type ReadWriteLock <: _Lock
+# Need Julia VERSION > v"1.2.0-DEV.28` to have `ReentrantLock <: AbstractLock`
+LockTypes = Union{AbstractLock, ReentrantLock}
```
Comment (nickrobinson251) [inline src/ReadWriteLocks.jl:21]: should this also include `Base.AlwaysLockedST`?

File: test/singlethreaded.jl (hunk 4 above)
Comment (nickrobinson251) [reply test/singlethreaded.jl]: https://github.com/invenia/ReadWriteLocks.jl/pull/4/commits/b006f13c1c58bee2dbc0c5bd5734367a33f88bc6

File: src/ReadWriteLocks.jl (hunk 6 above)
Comment (iamed2) [reply src/ReadWriteLocks.jl:21]: No, I don't think that's intended for public use

File: test/singlethreaded.jl (hunk 4 above)
Comment (iamed2) [reply test/singlethreaded.jl]: lol okay put them back I guess? Bizarre.

Comment (nickrobinson251) [reply test/singlethreaded.jl]: reverted 🤷‍♂

File: src/ReadWriteLocks.jl (hunk 2 above)
Comment (nickrobinson251) [reply src/ReadWriteLocks.jl:47]: You mean also rename `lock` -> `lock!` (even though we mutate the `rwlock.readers`)?

https://github.com/invenia/ReadWriteLocks.jl/issues/5

File: test/runtests.jl (hunk 5 above)
Comment (nickrobinson251) [reply test/runtests.jl]: in case we do this in a follow-up PR, opened https://github.com/invenia/ReadWriteLocks.jl/issues/6

**Review by iamed2 — APPROVED** [review]

**Pull request merged**